pyinjective/core/network.py

```diff
@@ -1,171 +1,103 @@
-            async with self._chain_cookie_initialization_lock:
-                if self._chain_cookie is None:
-                    await self._fetch_chain_cookie(metadata_query_provider=metadata_query_provider)
-        cookie = self._chain_cookie
-        self._check_chain_cookie_expiration()
+        # The cookies for these nodes do not expire
+        return False
 
-        return cookie
 
-    async def exchange_cookie(self, metadata_query_provider: Callable) -> str:
-        if self._exchange_cookie is None:
-            async with self._exchange_cookie_initialization_lock:
-                if self._exchange_cookie is None:
-                    await self._fetch_exchange_cookie(metadata_query_provider=metadata_query_provider)
-        cookie = self._exchange_cookie
-        self._check_exchange_cookie_expiration()
+class ExpiringCookieAssistant(CookieAssistant):
+    def __init__(self, expiration_time_keys_sequence: List[str], time_format: str):
+        self._cookie: Optional[str] = None
+        self._expiration_time_keys_sequence = expiration_time_keys_sequence
+        self._time_format = time_format
 
-        return cookie
+    @classmethod
+    def for_kubernetes_public_server(cls):
+        return cls(
+            expiration_time_keys_sequence=["grpc-cookie", "expires"],
+            time_format="%a, %d-%b-%Y %H:%M:%S %Z",
+        )
 
-    async def _fetch_chain_cookie(self, metadata_query_provider: Callable):
-        metadata = await metadata_query_provider()
-        cookie_info = next((value for key, value in metadata if key == "set-cookie"), None)
+    def cookie(self) -> Optional[str]:
+        self._check_cookie_expiration()
 
-        if cookie_info is None:
-            raise RuntimeError(f"Error fetching chain cookie ({metadata})")
+        return self._cookie
 
-        self._chain_cookie = cookie_info
+    async def process_response_metadata(self, grpc_call: Call):
+        metadata = await grpc_call.initial_metadata()
+        if "set-cookie" in metadata:
+            self._cookie = metadata["set-cookie"]
 
-    async def _fetch_exchange_cookie(self, metadata_query_provider: Callable):
-        metadata = await metadata_query_provider()
-        cookie_info = next((value for key, value in metadata if key == "set-cookie"), None)
+    def _check_cookie_expiration(self):
+        if self._is_cookie_expired():
+            self._cookie = None
 
-        if cookie_info is None:
-            cookie_info = ""
+    def _is_cookie_expired(self) -> bool:
+        is_expired = False
 
-        self._exchange_cookie = cookie_info
+        if self._cookie is not None:
+            cookie = SimpleCookie()
+            cookie.load(self._cookie)
+            cookie_map = cookie
 
-    def _check_chain_cookie_expiration(self):
-        if self._is_cookie_expired(cookie_data=self._chain_cookie):
-            self._chain_cookie = None
+            for key in self._expiration_time_keys_sequence[:-1]:
```

Comment:
Why are you skipping last item in this iteration?

Reply:
The variable `_expiration_time_keys_sequence` has all the keys that should be used to access the cookie in sequence (the cookie can be in a nested dictionary, to explain it better). So we use the list of keys without the last element just to keep going into the nested dictionaries (like a Mamushka of dictionaries), and the last element to access the cookie value